File: samples/client/petstore/csharp/src/main/csharp/io/swagger/client/ApiClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using Newtonsoft.Json;
using RestSharp;

namespace IO.Swagger.Client {
  /// <summary>
  /// API client is mainly responible for making the HTTP call to the API backend
  /// </summary>
  public class ApiClient {

    /// <summary>
    /// Initializes a new instance of the <see cref="ApiClient"/> class.
    /// </summary>
    /// <param name="basePath">The base path.</param>
    public ApiClient(String basePath="http://petstore.swagger.io/v2") {
      this.basePath = basePath;
      this.restClient = new RestClient(this.basePath);
    }

    /// <summary>
    /// Gets or sets the base path.
    /// </summary>
    /// <value>The base path.</value> 
    public string basePath { get; set; }

    /// <summary>
    /// Gets or sets the RestClient
    /// </summary>
    /// <value>The RestClient.</value> 
    public RestClient restClient { get; set; }

    private Dictionary<String, String> defaultHeaderMap = new Dictionary<String, String>();

    public Object CallApi(String Path, RestSharp.Method Method, Dictionary<String, String> QueryParams, String PostBody, 
      Dictionary<String, String> HeaderParams, Dictionary<String, String> FormParams, Dictionary<String, String> FileParams, String[] AuthSettings) {

      var request = new RestRequest(Path, Method);

      UpdateParamsForAuth(QueryParams, HeaderParams, AuthSettings);

      // add default header, if any
      foreach(KeyValuePair<string, string> defaultHeader in this.defaultHeaderMap)
        request.AddHeader(defaultHeader.Key, defaultHeader.Value);

      // add header parameter, if any
      foreach(KeyValuePair<string, string> param in HeaderParams)
        request.AddHeader(param.Key, param.Value);
     
      // add query parameter, if any
      foreach(KeyValuePair<string, string> param in QueryParams)
        request.AddQueryParameter(param.Key, param.Value);

      // add form parameter, if any
      foreach(KeyValuePair<string, string> param in FormParams)
        request.AddParameter(param.Key, param.Value);

      // add file parameter, if any
      foreach(KeyValuePair<string, string> param in FileParams)
        request.AddFile(param.Key, param.Value);

      if (PostBody != null) {
        request.AddParameter("application/json", PostBody, ParameterType.RequestBody); // http body (model) parameter
      }

      return (Object)restClient.Execute(request);

    }

    /// <summary>
    /// Add default header
    /// </summary>
    /// <param name="key">  Header field name
    /// <param name="value"> Header field value
    /// <returns></returns>
    public void AddDefaultHeader(string key, string value) {
       defaultHeaderMap.Add(key, value);
    }

    /// <summary>
    /// Get default header
    /// </summary>
    /// <returns>Dictionary of default header</returns>
    public Dictionary<String, String> GetDefaultHeader() {
       return defaultHeaderMap;
    }

    /// <summary>
    /// escape string (url-encoded)
    /// </summary>
    /// <param name="str"> String to be escaped
    /// <returns>Escaped string</returns>
    public string EscapeString(string str) {
      return str;
    }

    /// <summary>
    /// if parameter is DateTime, output in ISO8601 format
    /// if parameter is a list of string, join the list with ","
    /// otherwise just return the string
    /// </summary>
    /// <param name="obj"> The parameter (header, path, query, form)
    /// <returns>Formatted string</returns>
    public string ParameterToString(object obj)
    {
      if (obj is DateTime) {
        return ((DateTime)obj).ToString ("u");
      } else if (obj is List<string>) {
        return String.Join(",", obj as List<string>);
      } else {
	return Convert.ToString (obj);
      }
    }

    /// <summary>
    /// Deserialize the JSON string into a proper object
    /// </summary>
    /// <param name="json"> JSON string
    /// <param name="type"> Object type
    /// <returns>Object representation of the JSON string</returns>
    public object Deserialize(string content, Type type) {
      if (type.GetType() == typeof(Object))
        return (Object)content;

      try
      {
          return JsonConvert.DeserializeObject(content, type);
      }
      catch (IOException e) {
        throw new ApiException(500, e.Message);
      }
    }

    /// <summary>
    /// Serialize an object into JSON string
    /// </summary>
    /// <param name="obj"> Object 
    /// <returns>JSON string</returns>
    public string Serialize(object obj) {
      try
      {
          return obj != null ? JsonConvert.SerializeObject(obj) : null;
      }
      catch (Exception e) {
        throw new ApiException(500, e.Message);
      }
    }

    /// <summary>
    /// Get the API key with prefix
    /// </summary>
    /// <param name="obj"> Object 
    /// <returns>API key with prefix</returns>
    public string GetApiKeyWithPrefix (string apiKey)
    {
      var apiKeyValue = "";
      Configuration.apiKey.TryGetValue (apiKey, out apiKeyValue);
      var apiKeyPrefix = "";
      if (Configuration.apiKeyPrefix.TryGetValue (apiKey, out apiKeyPrefix)) {
        return apiKeyPrefix + " " + apiKeyValue;
      } else {
        return apiKeyValue;
      }
    }

    /// <summary>
    /// Update parameters based on authentication
    /// </summary>
    /// <param name="QueryParams">Query parameters</param>
    /// <param name="HeaderParams">Header parameters</param>
    /// <param name="AuthSettings">Authentication settings</param>
    public void UpdateParamsForAuth(Dictionary<String, String> QueryParams, Dictionary<String, String> HeaderParams, string[] AuthSettings) {
      if (AuthSettings == null || AuthSettings.Length == 0)
        return;
  
      foreach (string auth in AuthSettings) {
        // determine which one to use
        switch(auth) {
          
          case "api_key":
            HeaderParams["api_key"] = GetApiKeyWithPrefix("api_key");
            
            break;
          
          case "petstore_auth":
            
            //TODO support oauth
            break;
          
          default:
            //TODO show warning about security definition not found
          break;
        }
      }

    }

    /// <summary>
    /// Encode string in base64 format 
    /// </summary>
    /// <param name="text">String to be encoded</param>
    public static string Base64Encode(string text) {
      var textByte = System.Text.Encoding.UTF8.GetBytes(text);
      return System.Convert.ToBase64String(textByte);
    }

  }
}

File: samples/client/petstore/go/go-petstore/api_client.go
package swagger

import (
    "strings"
    "github.com/go-resty/resty"
    "errors"
    "reflect"
)

type ApiClient struct {

}

func (c *ApiClient) SelectHeaderContentType(contentTypes []string) string {
    if (len(contentTypes) == 0){
        return ""
    }
    if contains(contentTypes,"application/json") {
        return "application/json"
    }

    return contentTypes[0] // use the first content type specified in 'consumes'
}

func (c *ApiClient) SelectHeaderAccept(accepts []string) string {
    if (len(accepts) == 0){        
        return ""
    }

    if contains(accepts,"application/json"){        
        return "application/json"
    }

    return strings.Join(accepts,",")
}

func contains(source []string, containvalue string) bool {
    for _, a := range source {
        if strings.ToLower(a) == strings.ToLower(containvalue) {
            return true
        }
    }
    return false
}


func (c *ApiClient) CallApi(path string, method string,
    postBody interface{},
    headerParams map[string]string,
    queryParams map[string]string,
    formParams map[string]string,
    fileParams map[string]string) (*resty.Response, error) {

    request := prepareRequest(postBody, headerParams, queryParams, formParams, fileParams)

    switch strings.ToUpper(method) {
    case "GET":
        response, err := request.Get(path)
        return response, err
    case "POST":
        response, err := request.Post(path)
        return response, err
    case "PUT":
        response, err := request.Put(path)
        return response, err
    case "PATCH":
        response, err := request.Patch(path)
        return response, err
    case "DELETE":
        response, err := request.Delete(path)
        return response, err
    }

    return nil, errors.New("Invalid method " + method)
}

func (c *ApiClient) ParameterToString (obj interface{}) string {
    if reflect.TypeOf(obj).String() == "[]string" {
        return strings.Join(obj.([]string), ",")
    } else{
        return obj.(string)
    }
}

func prepareRequest(postBody interface{},
    headerParams map[string]string,
    queryParams map[string]string,
    formParams map[string]string,
    fileParams map[string]string) *resty.Request {

    request := resty.R()

    request.SetBody(postBody)

    // add header parameter, if any
    if len(headerParams) > 0 {
        request.SetHeaders(headerParams)
    }
    
    // add query parameter, if any
    if len(queryParams) > 0 {
        request.SetQueryParams(queryParams)
    }

    // add form parameter, if any
    if len(fileParams) > 0 {
        request.SetFormData(formParams)
    }

    // add file parameter, if any
    if len(fileParams) > 0 {
        request.SetFiles(fileParams)
    }
    
    return request
}